x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/registry/ModelRegistry.java

```diff
@@ -662,163 +661,261 @@ public void storeModel(Model model, ActionListener<Boolean> listener, TimeValue
+        var failure = item.getFailure();
+
+        String inferenceIdOrUnknown = "unknown";
+        var inferenceIdMaybeNull = docIdToInferenceId.get(item.getId());
+        if (inferenceIdMaybeNull == null) {
+            logger.warn("Failed to find inference id for document id [{}]", item.getId());
+        } else {
+            inferenceIdOrUnknown = inferenceIdMaybeNull;
+        }
+
+        if (item.isFailed() && failure != null) {
             logger.warn(
-                format("Failed to store inference endpoint [%s] index: [%s]", inferenceEntityId, item.getIndex()),
-                item.getFailure().getCause()
+                format(
+                    "Failed to store document id: [%s] inference id: [%s] index: [%s] bulk failure message [%s]",
+                    item.getId(),
+                    inferenceIdOrUnknown,
+                    item.getIndex(),
+                    item.getFailureMessage()
+                )
             );
+
+            return new ModelStoreResponse(inferenceIdOrUnknown, item.status(), failure.getCause());
+        } else {
+            return new ModelStoreResponse(inferenceIdOrUnknown, item.status(), null);
         }
     }
 
-    private static BulkItemResponse.Failure getFirstBulkFailure(BulkResponse bulkResponse) {
-        for (BulkItemResponse item : bulkResponse.getItems()) {
-            if (item.isFailed()) {
-                return item.getFailure();
-            }
+    private static Model getModelFromMap(@Nullable String inferenceId, Map<String, Model> inferenceIdToModel) {
+        if (inferenceId != null) {
+            return inferenceIdToModel.get(inferenceId);
         }
 
         return null;
     }
 
+    private void updateClusterState(List<Model> models, ActionListener<AcknowledgedResponse> listener, TimeValue timeout) {
+        var inferenceIdsSet = models.stream().map(Model::getInferenceEntityId).collect(Collectors.toSet());
+        var storeListener = listener.delegateResponse((delegate, exc) -> {
+            logger.warn(format("Failed to add inference endpoint %s minimal service settings to cluster state", inferenceIdsSet), exc);
+            deleteModels(
+                inferenceIdsSet,
+                ActionListener.running(
+                    () -> delegate.onFailure(
+                        new ElasticsearchStatusException(
+                            format(
+                                "Failed to add the inference endpoints %s. The service may be in an "
+                                    + "inconsistent state. Please try deleting and re-adding the endpoint.",
+                                inferenceIdsSet
+                            ),
+                            RestStatus.INTERNAL_SERVER_ERROR,
+                            exc
+                        )
+                    )
+                )
+            );
+        });
+
+        try {
+            metadataTaskQueue.submitTask(
+                format("add models %s", inferenceIdsSet),
```

Comment:
Might this be better as "add model metadata for %s"? We're not actually adding models here, and the inference ID is not strictly the same as a model.